File: old/Python/Driver.py
###############################################################################
#                                                                             #
#  Driver                                                                     #
#                                                                             #
#  This component is the entry point for the rest of the code.                #
#                                                                             #
###############################################################################

# General Python stuff
import argparse
import ConfigParser
import math
import os

# Imports specific to this code
# - I use "import x as x" for future development, when there may be more than
#   one implementation of the various components, so that you can do, for
#   example, "import GammaLaw as EoS" and the rest of the code will be written
#   generically in terms of EoS.
import Grid as Grid
import Hydro as Hydro
import IC as IC

# =============================================================================

MAX_ITER  = None
output_dt = None
time      = None
tmax      = None

# =============================================================================

def parse_command_line():
   """
   Parse the command-line arguments.
   """

   desc = "Perform fluid dynamics simulations."
   parser = argparse.ArgumentParser(description=desc)

   # Parameter file
   help_txt = "name of the configuration file (default is 'config.ini.')"
   parser.add_argument("-f", "--file", metavar="FILE", default="config.ini",
                       required=False, dest="config_file", help=help_txt)

   return parser.parse_args()

# =============================================================================

def parse_config_file(config_file):
   """
   Parse the configuration file.
   """

   config = ConfigParser.SafeConfigParser()
   config.read(config_file)

   config_dict = {}
   for section in config.sections():
      # TODO : Should I force all section names to lowercase?
      config_dict[section.strip()] = dict(config.items(section))


   return config_dict

# =============================================================================

def init():
   """
   Handle the initialization for the Driver component.
   """

   global MAX_ITER
   global output_dt
   global time
   global tmax

   # ==========================================================================
   # Arguments and parameters

   # Read the command-line arguments
   cline_args = parse_command_line()

   # Read the parameter file
   cfile_args = parse_config_file(cline_args.config_file)

   # Add the command-line arguments as section 'CommandLine'
   # TODO --- test this
   cfile_args['CommandLine'] = vars(cline_args)

   # ==========================================================================
   # Initialize the Driver component

   if 'Driver' not in cfile_args:
      cfile_args['Driver'] = {}

   # Maximum number of iterations
   if 'max_iter' in cfile_args['Driver']:
      MAX_ITER = int(cfile_args['Driver']['max_iter'])
   else:
      MAX_ITER = 10
   # Store either the default value or the value from the config file converted
   # from string to integer
   cfile_args['Driver']['max_iter'] = MAX_ITER

   # Time step between outputs
   if 'output_dt' in cfile_args['Driver']:
      output_dt = float(cfile_args['Driver']['output_dt'])
   else:
      output_dt = 0.0
   cfile_args['Driver']['output_dt'] = output_dt

   # Current time
   time = 0.0

   # Final time
   if 'tmax' in cfile_args['Driver']:
      tmax = float(cfile_args['Driver']['tmax'])
   else:
      raise Exception("tmax required") #TODO

   # ==========================================================================
   # Initialize the other components
   # - Each initialization updates its own parameters dictionary (adding
   #   default values for anything not specified on the command line or in the
   #   configuration file) and returns the updated dictionary for logging.

   # Initialize the grid
   if "Grid" not in cfile_args:
      cfile_args['Grid'] = {}
   cfile_args['Grid'] = Grid.init(cfile_args['Grid'])

   # Initialize the hydro
   if "Hydro" not in cfile_args:
      cfile_args['Hydro'] = {}
   cfile_args['Hydro'] = Hydro.init(cfile_args['Hydro'])

   # ==========================================================================
   # Write parameters to log file

   if not os.path.exists("output"):
      os.makedirs("output")

   f = open("output/parameters.txt",'w')
   for section in cfile_args:
      f.write("[ " + section + " ]\n")
      for item in cfile_args[section]:
         f.write("   " + item + " : " + str(cfile_args[section][item]) + "\n")
   f.close()

   # ==========================================================================
   # Set up initial conditions
   IC.set_initial_conditions()

   return cfile_args

# =============================================================================

def finalize():

   Grid.finalize()
   Hydro.finalize()

# =============================================================================

def compute_time_step():
   """
   Compute the time step.

   This may depend on multiple components, so it is packaged as a Driver
   function even though the current version only cares about the Hydro
   component.
   """

   dt = Hydro.compute_time_step()

   return dt

# =============================================================================

def main():
   """
   The main core of the program.
   """

   global time
   global MAX_ITER
   global output_dt

   # ==========================================================================
   # Initialize
   # - Returns a dictionary of dictionaries, which are the parameters from the
   #   configuration file (by section) and from the command line (listed as
   #   section 'CommandLine')
   params = init()

   # ==========================================================================
   # Evolution loop

   prev_write = -1

   for n_iter in xrange(MAX_ITER):

      # Exceeded maximum time
      if time >= tmax:
         break

      # Unexpected exit? (e.g. like "touch .dump_restart")
      # TODO

      # Write output
      do_write = False
      if output_dt > 0:
         curr_write = int(math.floor(time / output_dt))
         if curr_write > prev_write:
            do_write = True
         prev_write = curr_write
      if do_write:   # to allow for a condition based on step number
         Grid.write_data(n_iter)

      # Boundary condition fill
      Grid.fill_boundary_conditions()

      # Compute step size
      dt = compute_time_step()

      # Evolve a single step
      Hydro.one_step(dt)

      # Update time
      time = time + dt

   # End evolution loop

   # ==========================================================================
   # Finalize

   # Write final output
   Grid.write_data(n_iter+1)
   finalize()

File: 000_LinearAdvection_1D_Serial/Driver.cpp
// TODO : Do I want to turn this into a class to allow public/private?  There
// are also unnamed namespaces, but I don't know the usage of those so I'm not
// yet confident enough to use them myself.

// STL includes
#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

// Boost includes
#include <boost/filesystem.hpp>
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/ini_parser.hpp>

// Includes specific to this code
#include "Driver.hpp"
#include "Grid.hpp"
#include "Hydro.hpp"
#include "InitConds.hpp"
#include "Monitor.hpp"

namespace fs = boost::filesystem;
namespace pt = boost::property_tree;

namespace Driver {

   // parameter listings (all are case-insensitive PropertyTrees from Boost)
   pt::iptree param_used;     // parameters used by the code
   pt::iptree param_unused;   // parameters not used by the code
   pt::iptree param_changed;  // have parameters changed from default?

   // component-scope variables
   unsigned int n_step = 0;
   unsigned int max_steps = 0;
   unsigned int n_width;

   double dt;
   double time = 0.0;
   double tmax = 0.0;

   double output_dt = 0.0;

   std::string output_dir;
   std::string restart_dir;
   std::string log_file;
   std::ofstream lout;

   // =========================================================================
   // Parameter printing

   // Print the parameters from a section that has been processed
   // - Inputs:
   //   - name of the section to print
   // - Outputs:
   //   - none
   // - Side Effects:
   //   - none
   void print_parameters(std::string section_name) {
      pt::iptree used    = param_used.get_child(section_name);
      pt::iptree changed = param_changed.get_child(section_name);
      std::string key, value;
      int key_len, value_len;

      // Print the section name
      lout << std::endl << "[ " << section_name << " ]" << std::endl;

      // Sort the parameters by name
      used.sort();

      // Compute the lengths of the longest key and longest value (as string)
      key_len = 0;
      value_len = 0;
      for (pt::iptree::iterator pos = used.begin(); pos != used.end(); pos++) {
         key   = pos->first;
         value = used.get<std::string>(key);
         if (key.length() > key_len) {
            key_len = key.length();
         }
         if (value.length() > value_len) {
            value_len = value.length();
         }
      }

      // Print all the parameters
      for (pt::iptree::iterator pos = used.begin(); pos != used.end(); pos++) {
         key   = pos->first;
         value = used.get<std::string>(key);
         lout << "   " << std::left << std::setw(key_len) << key;
         lout << " : " << std::left << std::setw(value_len) << value;
         // Flag any that have been changed from default
         if (changed.get<bool>(key)) {
            lout << "   (CHANGED)";
         }
         lout << std::endl;
      }
   }

   // Print the parameters from the configuration file that were not used
   // - Inputs:
   //   - none
   // - Outputs:
   //   - none
   // - Side Effects:
   //   - none
   void print_unused_parameters() {
      
      pt::iptree::iterator outer, inner;
      pt::iptree child;

      // Don't do anything if the tree is empty
      if (!param_unused.empty()) {
         lout << std::endl << "Unused Parameters:" << std::endl;

         // Loop over all the sections in sorted order
         param_unused.sort();
         for (outer = param_unused.begin();
               outer != param_unused.end(); outer++) {

            // Get the subtree of all parameters in this section
            child = param_unused.get_child(outer->first);

            // Don't do anything if the subtree is empty
            if (!child.empty()) {

               // Print the section header
               lout << std::endl;
               lout << "[ " << outer->first << " ]" << std::endl;

               // Loop over all parameters in sorted order and print
               // - format is "   key : value" or "   key : value   (CHANGED)"
               child.sort();
               for (inner = child.begin(); inner != child.end(); inner++) {
                  lout << "   " << inner->first;
                  lout << " : " << child.get<std::string>(inner->first);
                  lout << std::endl;
               }
            }
         }
      }

      lout << std::endl;
   }

   // =========================================================================
   // Set up

   void setup (unsigned int argc, char *argv[]) {

      // ----------------------------------------------------------------------
      // Declare variables
      std::string param_file_name;
      bool changed_pf_name;

      // ----------------------------------------------------------------------
      // Arguments and parameters

      // Read the command-line arguments
      if (argc > 1) {
         param_file_name = argv[1];
         changed_pf_name = true;
      } else {
         param_file_name = "params.ini";
         changed_pf_name = false;
      }

      // Read the parameter file
      try {
         pt::read_ini(param_file_name, param_unused);
      } catch (pt::ini_parser_error &e) {
         std::cerr << "ERROR: Could not parse parameter file \"";
         std::cerr << param_file_name << "\"." << std::endl;
         throw;
      }

      // Add the command-line arguments to the parameter set
      param_used.put("Driver.config_file", param_file_name);
      param_changed.put("Driver.config_file", changed_pf_name);

      // ----------------------------------------------------------------------
      // Initialize the Driver component

      // Maximum number of time steps
      max_steps =
         parameter_with_default<unsigned int>("Driver.max_steps", 1000000);
      n_width = floor(log10(max_steps)) + 1;

      // Output directory
      output_dir =
         parameter_with_default<std::string>("Driver.output_dir", "output/");
      // Modify the output directory name as needed
      if (!output_dir.empty()) {
         if (*output_dir.rbegin() != '/') {
            output_dir.append("/");
         }
      } else {
         output_dir = "./";
      }
      // Ensure the output directory exists
      if (!fs::exists(output_dir)) {
         fs::create_directories(output_dir);
      }

      // Name of log file
      log_file =
         parameter_with_default<std::string>("Driver.log_file", "log.txt");
      log_file = output_dir + log_file;
      lout.open(log_file.c_str());

      // Time between saving output files
      output_dt = parameter_with_default<double>("Driver.output_dt", 0.0);

      // Restart from which directory
      restart_dir =
         parameter_with_default<std::string>("Driver.restart_dir", "");

      // Current time
      time = 0.0;

      // Final time
      tmax = parameter_without_default<double>("Driver.tmax");

      // Print the parameters for the Driver component
      lout << std::string(79, '_') << std::endl;
      lout << "Parameters:" << std::endl;

      print_parameters("Driver");

      // ----------------------------------------------------------------------
      // Initialize other components

      // Initialize the Grid component
      Grid::setup();
      print_parameters("Grid");

      // Initialize the Hydro component
      Hydro::setup();
      print_parameters("Hydro");

      // Initialize the InitConds (initial conditions) component
      InitConds::setup();
      print_parameters("InitConds");

      // Initialize the Monitor component
      Monitor::setup();
      print_parameters("Monitor");

      // Print the unused parameters
      print_unused_parameters();
   }

   // =========================================================================
   // Clean up

   void cleanup () {
      // Have all the other sections run their clean-up routines
      // --> Reverse order from setup in case of dependencies
      Monitor::cleanup();
      InitConds::cleanup();
      Hydro::cleanup();
      Grid::cleanup();
      // Close the log file
      lout << std::endl << std::string(79, '_') << std::endl;
      lout << "Program complete" << std::endl;
      lout.close();
   }

   // =========================================================================
   // Compute the time step
   // --> Each component may have restrictions on the time step; this collects
   //     all the time step restrictions and determines the maximum allowed
   //     time step consistent with all those restrictions.

   double compute_time_step() {
      double dt;
      double hydro_dt = Hydro::compute_time_step();
      dt = hydro_dt;
      /* For other components:
      double ???_dt = ???::compute_time_step();
      if (???_dt < dt) {
         dt = ???_dt;
      }
      */
      return dt;
   }

   // =========================================================================
   // The main evolution loop

   void main (unsigned int argc, char *argv[]) {

      // ----------------------------------------------------------------------
      // Declare variables

      // Tracking output
      double prev_write = -1;
      double curr_write;
      bool do_write;
      std::string outname;

      // ----------------------------------------------------------------------
      // Initialize

      setup(argc, argv);

      // ----------------------------------------------------------------------
      // Evolution loop

      lout << std::string(79, '_') << std::endl;
      lout << "Evolution loop:" << std::endl;

      for (; n_step < max_steps; n_step++) {

         // Exceeded maximum time
         if (time >= tmax) {
            break;
         }

         // Unexpected exit (e.g., like FLASH's "touch .dump_restart")
         // TODO

         // Write output
         do_write = false;
         // Condition based on time (print every output_dt seconds)
         if ((time == 0) || (output_dt > 0)) {
            curr_write = floor(time / output_dt);
            if (curr_write > prev_write) {
               do_write = true;
            }
            prev_write = curr_write;
         }
         // Conditions based on steps (print every output_steps seconds)
         // TODO
         // Do the actual write
         if (do_write) {
            outname = Grid::write_data();
            lout << "OUTPUT : wrote output " << outname << std::endl;
         }

         // Monitor write
         Monitor::write_to_monitor();

         // Boundary condition fill
         Grid::fill_boundary_conditions();

         // Compute step size
         dt = compute_time_step();

         // Print logfile note
         lout << "n = " << std::setw(n_width) << std::right << n_step;
         lout << "; t = " << std::setw(13) << std::scientific << time;
         lout << "; dt = " << std::setw(13) << std::scientific << dt;
         lout << std::endl;

         // Evolve a single step
         Hydro::one_step();

         // Update time
         time = time + dt;

      }

      // Final write
      lout << "n = " << std::setw(n_width) << std::right << n_step;
      lout << "; t = " << std::setw(13) << std::scientific << time;
      lout << std::endl;
      outname = Grid::write_data();
      lout << "OUTPUT : wrote output \"" << outname << "\"" << std::endl;
      Monitor::write_to_monitor();

      // ----------------------------------------------------------------------
      // Finalize

      cleanup();

   }

}
